### ISSUE TITLE
musepack sv7 files miss first 0.8 s when played

### ISSUE THREAD

**Issue by wutschel** [xbmc/xbmc#15945]
### Describe the bug
When playing audio files in the Musepack SV7 format (file ending ".mpc") the first 0.8 s are never played.

## Expected Behavior
The first 0.8 s of Musepack SV7 files should be played as well.

## Actual Behavior
Currently the playback in Kodi is always skipping the first 32 frames (0.8 s) of each Musepack SV7 file. This happens since Kodi 15.x and is present in all versions since then. I also verfiied this issue in Leia 18.1 and 18.2 RC1.
Reason is that the function "mpc7_decode_flush" must only be called after a seek to a frame position > 32. When this function is called the next 32 frames are muted. In Leia 18.2 this function is called twice at the beginning of each track. In consequence the first 32 frames of each track are missing.

## Possible Fix
As a fix it is possible to check if the flush function is called for the first decoded frame or not. If so, the flush-processing must be skipped. The code is located in the ffmpeg source code under libavcodec/mpc7.c: https://pastebin.com/xzG3Am43

### To Reproduce
The bug can be best reproduced with tracks that contain significant audio signal in  the first second. Example: https://www.sendspace.com/file/xju086. 
When playing this track the initial breathing-in of the singer should be audible. Using Kodi this breath is not present, using other programs (e.g. ffmpeg, foobar2000) the file is properly playbed back.

### Debuglog
You will not see the issue from the logs. I needed to add new log messages to be able to debug this.

### Screenshots 
Not needed.

## Additional context or screenshots (if appropriate)
Please also see the conversion on kodi.tv: https://forum.kodi.tv/showthread.php?tid=330109

### Your Environment
Used Operating system:
 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Linux Mint 19.1
 - Kodi version: 18.2 RC1

**Comment by fritsch** [comment]
I have read your forum post and seen that you adjust ffmpeg. Please submit your patch if you think this needs to be adjusted in ffmpeg to: http://ffmpeg.org/pipermail/ffmpeg-devel/ mailing list.

**Comment by wutschel** [comment]
I am not sure if this needs to be applied (or even can be applied the same way) to ffmpeg in general. Other players using ffmpeg do not show the issue which is observed with Kodi. Doesn´t it make sense to apply the patch to Kodi´s local ffmpeg version?

**Comment by fritsch** [comment]
No, it makes no sense at all. If we use ffmpeg wrong, we need to fix it. If it's wrong upstream it needs fixing there.

Adding custom patches into an upstream library will just increase kodi's maintenance costs and make kodi less compatible.

**Comment by wutschel** [comment]
Understood. So there are two options left:
1) Find out why Kodi is calling flush() without need.
2) Find out if the ffmpeg team would accept the code change for the robustness in general.

**Comment by fritsch** [comment]
Jep. I think 1 is preferred :-)

**Comment by wutschel** [comment]
Good. Then let me add a log which I took using the patch I posted here as possible solution. When searching for the keyword "mpc7_decode_flush" you will find several occurances which show "frame #: 1". These are the unexpected double calls at the start of each file. The other occurances with frame numbers !=1 only show up once after triggering a seek within a file, which is the expected and proper behaviour.
Log: https://pastebin.com/LPZDHRRS

**Comment by fritsch** [comment]
Okay, I think I know what happens:
```
2019-04-20 22:24:21.972 T:140184458635008   DEBUG: SeekTime - unknown position after seek
```

After that we seek, which causes the Flush.

Can you check: bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double *startpts) especially time and *startpts ?

```
  // in this case the start time is requested time
  if (startpts)
    *startpts = DVD_MSEC_TO_TIME(time);
```

**Comment by fritsch** [comment]
I think there is already a workaround for mp3 - let's see what you find by spankling debug outputs in this function.

**Comment by wutschel** [comment]
In the error case SeekTime() is called like this:
time: 1 startpts: NULL
-> flush
time: 0 startpts: NULL
-> flush

In the good case (seek) it look like this:
time: 69847: startpts: NULL
-> flush

**Comment by wutschel** [comment]
Found the root cause.

DVDDemuxFFmpeg::Seektime() is called from VideoPlayerCodec::Seek(). In VideoPlayerCodec:Init() there is a section which seems to check if the codec has seek-capabilty. The check performs seek to 1 and back to 0:

> if (Seek(1))
> {
>   // rewind stream to beginning
>   Seek(0);
>   m_bCanSeek = true;
> }
> else
> {
>  ...
> }

Replacing the whole block with

> m_bCanSeek = true;

fixes the behaviour. No uneeded flush() calls anymore.

**Comment by fritsch** [comment]
seek_pts += m_pFormatContext->start_time; <- is this changed on first call as well?

**Comment by fritsch** [comment]
@wutschel Okay. Thx much.

@FernetMenta do you see a possibile issue in the SeekBack code?

**Comment by FernetMenta** [comment]
imo a clear ffmpeg issue. skipping the next 32 frames after flush is incredible nonsense. the author of that code must have been drunken :)

**Comment by wutschel** [comment]
True, one can argue that ffmpeg is misusing the flush() command to handle some seek-specific behaviour which should better be handled at another place. But also Kodi´s way of calling seek() multiple times (each time followed by flush()), for each new tack and just for the sake of checking if seeking works, is not well designed. I am -- despite any discussion about ffmpeg -- still in favour to change the implementation inside VideoPlayerCodec:Init().

**Comment by FernetMenta** [comment]
The root cause can't be fixed in Kodi. The ffmpeg bug won't let you seek to the very beginning of a track without missing a couple of frames.

**Comment by wutschel** [comment]
I agree to your point of where the root cause is. Nevertheless the player should not seek to 1 s, followed by seek to 0 s for each new track. And this is an issue of kodi and not ffmpeg. If this would be fixed, the bug in ffmpeg would be irrelevant as it would not be triggered anymore during normal playback. I say "irrelevant" as skipping back or forward will always start a new track (no seek to 0 s is performed), and as a seek to a timestamp <1 s is hard to trigger via UI.

**Comment by FernetMenta** [comment]
Feel free to implement another method of testing whether seek is possible on a stream. Your suggested approach above is obviously not a solution.

**Comment by fritsch** [comment]
Sledge-Hammer:

```
@@ -214,17 +214,11 @@ bool VideoPlayerCodec::Init(const CFileItem &file, unsigned int filecache)
   if (m_pInputStream->Seek(0, SEEK_POSSIBLE))
   {
     if (Seek(1))
-    {
-      // rewind stream to beginning
-      Seek(0);
       m_bCanSeek = true;
-    }
-    else
-    {
-      m_pInputStream->Seek(0, SEEK_SET);
-      if (!m_pDemuxer->Reset())
-        return false;
-    }
+
+    m_pInputStream->Seek(0, SEEK_SET);
+    if (!m_pDemuxer->Reset())
+      return false;
   }
```

With all the disadvantages it brings.

**Comment by wutschel** [comment]
Isn´t

`if (m_pInputStream->Seek(0, SEEK_POSSIBLE)`

already checking if a seek is possible in the stream? If this is not the case, then I can understand the patch is not an acceptable solution. In this case I will stick to my local patch of ffmpeg and keep my local build of kodi.

**Comment by FernetMenta** [comment]
> Isn´t
if (m_pInputStream->Seek(0, SEEK_POSSIBLE)
already checking if a seek is possible in the stream?

So you think the current code was added just for fun or to make things worse? Using git blame brings you to this change: https://github.com/xbmc/xbmc/pull/4594

inputStream != demuxer, in other words: being able to set the file pos on an input stream does not mean that the demuxer has implemented a seek function.

You can sacrifice one scenario in favour of another one, or you can fix the root cause. Personlly I don't care. I am just providing the infos that help to fix it properly.

**Comment by wutschel** [comment]
Next option: As check for seek capability replace calling VideoPlayerCodec::Seek() with demuxer´s SeekTime(). This way we do not call m_pAudioCodec->Reset(). This also avoids calling flush and still allows to probe the seek support of the demuxer.

```
@@ -213,10 +213,10 @@ bool VideoPlayerCodec::Init(const CFileItem &file, unsigned int filecache)
   m_bCanSeek = false;
   if (m_pInputStream->Seek(0, SEEK_POSSIBLE))
   {
-    if (Seek(1))
+    if (m_pDemuxer->SeekTime(1, true))
     {
       // rewind stream to beginning
-      Seek(0);
+      m_pDemuxer->SeekTime(0, true);
       m_bCanSeek = true;
     }
     else
```

**Comment by FernetMenta** [comment]
I don't see any negative impacts. Good idea.

**Comment by fritsch** [comment]
@wutschel ready to send a PullRequest? Thanks in advance.

**Comment by wutschel** [comment]
@FernetMenta: Thanks for your patience with me :)
@fritsch: I never did this in github/kodi. I would not mind, if you would create it, Otherwise I will need some time to read through the documentation and the kodi rules.

**Comment by fritsch** [comment]
@wutschel Thx very much. PR is sent. I would like to get it tested in our nightlies for now. If everything works out well, it might be a candidate for 18.3.

For the time being you can cherry-pick it to whatever version you are currently building.

**Comment by wutschel** [comment]
Great. Let me know if I can be of further assistance.

**Comment by wutschel** [comment]
Took me a while to get used the ffmpeg code base, but this finally fixes the seek correctly. It is not a workaround, it is a proper fix to ffmpeg which correctly support seeking to any position in the file. I am not fully clear on how to share this with ffmpeg, and I am also not sure, if this will make it. Just wanted to share this here, just in case someone might want to apply it.

```
diff --git a/libavcodec/mpc7.c b/libavcodec/mpc7.c
index 98ef1732cf..105319818b 100644
--- a/libavcodec/mpc7.c
+++ b/libavcodec/mpc7.c
@@ -42,6 +42,8 @@
 #include "mpc.h"
 #include "mpc7data.h"
 
+#define MPC_PRIMING_FRAMES 32
+
 static VLCElem scfi_vlc[1 << MPC7_SCFI_BITS];
 static VLCElem dscf_vlc[1 << MPC7_DSCF_BITS];
 static VLCElem hdr_vlc [1 << MPC7_HDR_BITS];
@@ -200,6 +202,18 @@ static int mpc7_decode_frame(AVCodecContext *avctx, AVFrame *frame,
         av_log(avctx, AV_LOG_WARNING, "packet size is not a multiple of 4. "
                "extra bytes at the end will be skipped.\n");
     }
+    
+    /* Packet layout (set by mpc demuxer):
+     * buf[0]: skip
+     * buf[1]: last_frame
+     * buf[2]: frames to skip after seek (range 0..MPC_PRIMING_FRAMES)
+     * buf[3]: unused
+     */
+    
+    /* buf[2] holds the number of frames to be skipped after a seek */
+    if (buf[2] > 0) {
+        c->frames_to_skip = FFMIN(buf[2], MPC_PRIMING_FRAMES);
+    }
 
     skip       = buf[0];
     last_frame = buf[1];
@@ -302,7 +316,6 @@ static av_cold void mpc7_decode_flush(AVCodecContext *avctx)
     MPCContext *c = avctx->priv_data;
 
     memset(c->oldDSCF, 0, sizeof(c->oldDSCF));
-    c->frames_to_skip = 32;
 }
 
 static av_cold int mpc7_decode_close(AVCodecContext *avctx)
diff --git a/libavformat/mpc.c b/libavformat/mpc.c
index 45ba76c703..838763ddab 100644
--- a/libavformat/mpc.c
+++ b/libavformat/mpc.c
@@ -45,6 +45,7 @@ typedef struct MPCContext {
     MPCFrame *frames;
     int curbits;
     int frames_noted;
+    int frames_to_skip;
 } MPCContext;
 
 static int mpc_probe(const AVProbeData *p)
@@ -164,6 +165,12 @@ static int mpc_read_packet(AVFormatContext *s, AVPacket *pkt)
     pkt->data[1] = (c->curframe > c->fcount) && c->fcount;
     pkt->data[2] = 0;
     pkt->data[3] = 0;
+    
+    /* After seek, share the frames to be skipped with the decoder */
+    if (c->frames_to_skip > 0) {
+        pkt->data[2] = FFMIN(c->frames_to_skip, DELAY_FRAMES);
+        c->frames_to_skip = 0;
+    }
 
     pkt->stream_index = 0;
     pkt->pts = cur;
@@ -191,21 +198,27 @@ static int mpc_read_seek(AVFormatContext *s, int stream_index, int64_t timestamp
     FFStream *const sti = ffstream(st);
     MPCContext *c = s->priv_data;
     AVPacket pkt1, *pkt = &pkt1;
-    int ret;
+    int ret, frameskip;
     int index = av_index_search_timestamp(st, FFMAX(timestamp - DELAY_FRAMES, 0), flags);
     uint32_t lastframe;
 
-    /* if found, seek there */
+    /* If found, seek there and update frames to skip */
     if (index >= 0 && sti->index_entries[sti->nb_index_entries-1].timestamp >= timestamp - DELAY_FRAMES) {
         c->curframe = sti->index_entries[index].pos;
+        c->frames_to_skip = FFMIN(timestamp, DELAY_FRAMES);
         return 0;
     }
     /* if timestamp is out of bounds, return error */
     if(timestamp < 0 || timestamp >= c->fcount)
         return -1;
-    timestamp -= DELAY_FRAMES;
+
+    /* When seeking to frame 0..31, only 0..31 frames must be skipped.
+     * For all other cases we must skip DELAY_FRAMES (= 32) frames */
+    frameskip = FFMIN(timestamp, DELAY_FRAMES);
+    timestamp = FFMAX(timestamp - DELAY_FRAMES, 0);
+
     /* seek to the furthest known position and read packets until
-       we reach desired position */
+     * we reach desired position, then set frames to skip */
     lastframe = c->curframe;
     if(c->frames_noted) c->curframe = c->frames_noted - 1;
     while(c->curframe < timestamp){
@@ -216,6 +229,7 @@ static int mpc_read_seek(AVFormatContext *s, int stream_index, int64_t timestamp
         }
         av_packet_unref(pkt);
     }
+    c->frames_to_skip = frameskip;
     return 0;
 }
```

**Comment by wutschel** [comment]
@fritsch, as this is still not fixed, can you re-open this one? The PR which was merged few years back was reverted duet regressions.

**Comment by fritsch** [comment]
Sure I can reopen. Issues shall be fixed upstream in ffmpeg. When this has happened a backport for internal ffmpeg of kodi can be considered,

**Comment by fritsch** [comment]
ffmpeg uses a mailing list to discuss patches. Here you are asked to post the patches, see:  https://github.com/FFmpeg/FFmpeg/blob/master/CONTRIBUTING.md

**Comment by garbear** [comment]
Is this still an issue?

**Comment by wutschel** [comment]
> Is this still an issue?

Yes, it is. I doubt the patch mentioned above will ever make it into Kodi though. But honestly, ffmpeg's mailing list way of handing things is too awkward.

**Comment by malard** [comment]
*This reply was written and posted by Claude on @malard's behalf.*

Removing `Resolution: Fixed`. It was wrong and it contradicted the `Triage: Needed` label sitting beside it.

@wutschel confirmed on 2026-06-29 that the first 0.8 s is still missing, and the position on the thread is unchanged: the defect is in ffmpeg's `mpc7` decoder, @fritsch's view is that it should be fixed upstream in ffmpeg with a backport considered afterwards, and nobody has taken the patch to the ffmpeg mailing list.

That is not "fixed" - it is a real, still-reproducible defect that is blocked on upstream. Marking it fixed risks it being closed on the strength of a label rather than on the state of the code.

Labelling `Component: Audio`.
